### ISSUE TITLE
Incompatible versions of scala-xml are causing problems with incremental migration to Scala3

### ISSUE THREAD

**Issue by liosedhel** [scalatest/scalatest#2092]
Hello! :)
After https://github.com/scalatest/scalatest/pull/1974 we have the following situation:
* for scalatest_3 - scala-xml version 2.0.0 is used
* for scalatest_2.13 - scala.xml version 1.3.0 is used

Those libraries are incompatible (see: https://github.com/scala/scala-xml/blob/v2.0.0-M1/CHANGELOG.md). Now, if some project uses in tests incompatible features from scala-xml 1.3.0 (like a reference to `adapter.hStack`), it is not possible to have the same module cross-compiled for Scala 2 and Scala 3 at the same time, which hinders the project's incremental migration to Scala 3.

Would you consider publishing a new scalatest_2.13 version also with the scala-xml 2.0.0 version? This will be a breaking change for some Scala 2 projects (as it is now with Scala 3 version), but at least the libraries will be aligned, and users will be able to cross compile for Scala 2 and Scala 3. 

Or is there a real problem which this different scala-xml versions (for scalatest_2.13 and scalatest_3 artifacts) solve and I am just missing a point? :)

**Comment by cheeseng** [comment]
@liosedhel Thanks for pointing this out, I have submitted the following PR to update scala-xml to 2.0.1 for build target that has scala-xml 2.x build: 

https://github.com/scalatest/scalatest/pull/2093

It shouldn't have with 2.13 and 3, it is just scala-xml 2.x is not published for earlier Scala version like 2.10 and 2.11, which still needs to be supported by ScalaTest 3.2.

MIMA seems happy with the change, I don't think we are using any in-compatible classes between scala-xml 1.3 and 2.x, so I think this should be safe.

For temporary workaround I wonder if you can exclude the transitive scala-xml 1.3.0 dependency and include scala-xml 2.0.1 instead?

**Comment by liosedhel** [comment]
Hi @cheeseng 
Thanks for a quick reply. Yes, that workaround I already applied so this is not a blocker at the moment, we can live with it for now :)

**Comment by limansky** [comment]
Hi,

  I've faced with scala-xml-2.x incompatibility with Scala 2.12. In our project we depend on scala-compiler, which depends on scala-xml-1.0.6. I've tried to upgrade to `ScalaTest` 3.2.11 and it leads to following error:
  
  ```
  sbt:yupana> benchmarks/dependencyTree
[error] stack trace is suppressed; run last benchmarks / dependencyTreeIgnoreMissingUpdate for the full output
[error] (benchmarks / dependencyTreeIgnoreMissingUpdate) found version conflict(s) in library dependencies; some are suspected to be binary incompatible:
[error] 
[error] 	* org.scala-lang.modules:scala-xml_2.12:2.0.1 (early-semver) is selected over 1.0.6
[error] 	   +- org.scalatest:scalatest-core_2.12:3.2.11           (depends on 2.0.1)
[error] 	   +- org.scala-lang:scala-compiler:2.12.15              (depends on 1.0.6)
```

For Scala 2.13 it works fine.

**Comment by cheeseng** [comment]
@limansky Yes in 3.2.11 we upgraded to scala-xml 2.x for Scala 2.12 and 2.13 build, afaik from ScalaTest side it should be compatible with both 1.x and 2.x scala-xml, so hopefully you may exclude the transitive scala-xml dependency and stay 1.x when required.

**Comment by limansky** [comment]
@cheeseng thanks for suggestion. I'll do it this way. But shouldn't ScalaTest for 2.12 depend on scala-xml 1.x?

**Comment by cheeseng** [comment]
@limansky scala-xml is de-coupled with Scala starting Scala 2.11, so I think it is good if we can move forward to use the new version by default.  Before this ticket it was the other way around where @liosedhel need to exclude the 1.x and include 2.x in his way to upgrade to Scala 3.  imho it is better we move to newer version by default and if really need the old 1.x there's still the exclude way.
